sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/MultimapUserStateTest.java

```diff
@@ -0,0 +1,385 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class MultimapUserStateTest {
+
+  private final String pTransformId = "pTransformId";
+  private final String stateId = "stateId";
+  private final String encodedKey = "encodedKey";
+  private final String encodedWindow = "encodedWindow";
+
+  @Test
+  public void testNoPersistedValues() throws Exception {
+    FakeBeamFnStateClient fakeClient = new FakeBeamFnStateClient(Collections.emptyMap());
+    MultimapUserState<String, String> userState =
+        new MultimapUserState<>(
+            fakeClient,
+            "instructionId",
+            pTransformId,
+            stateId,
+            encode(encodedWindow),
+            encode(encodedKey),
+            StringUtf8Coder.of(),
+            StringUtf8Coder.of());
+    assertTrue(Iterables.isEmpty(userState.keys()));
+  }
+
+  @Test
+  public void testGet() throws Exception {
+    FakeBeamFnStateClient fakeClient =
+        new FakeBeamFnStateClient(
+            ImmutableMap.of(
+                createMultimapKeyStateKey(),
+                encode("A1"),
+                createMultimapValueStateKey("A1"),
+                encode("V1", "V2")));
+    MultimapUserState<String, String> userState =
+        new MultimapUserState<>(
+            fakeClient,
+            "instructionId",
+            pTransformId,
+            stateId,
+            encode(encodedWindow),
+            encode(encodedKey),
+            StringUtf8Coder.of(),
+            StringUtf8Coder.of());
+
+    Iterable<String> initValues = userState.get("A1");
+    userState.put("A1", "V3");
+    assertArrayEquals(new String[] {"V1", "V2"}, Iterables.toArray(initValues, String.class));
+    assertArrayEquals(
+        new String[] {"V1", "V2", "V3"}, Iterables.toArray(userState.get("A1"), String.class));
+    assertArrayEquals(new String[] {}, Iterables.toArray(userState.get("A2"), String.class));
+    userState.asyncClose();
+    assertThrows(IllegalStateException.class, () -> userState.get("A1"));
+  }
+
+  @Test
+  public void testClear() throws Exception {
+    FakeBeamFnStateClient fakeClient =
+        new FakeBeamFnStateClient(
+            ImmutableMap.of(
+                createMultimapKeyStateKey(),
+                encode("A1"),
+                createMultimapValueStateKey("A1"),
+                encode("V1", "V2")));
+    MultimapUserState<String, String> userState =
+        new MultimapUserState<>(
+            fakeClient,
+            "instructionId",
+            pTransformId,
+            stateId,
+            encode(encodedWindow),
+            encode(encodedKey),
+            StringUtf8Coder.of(),
+            StringUtf8Coder.of());
+
+    Iterable<String> initValues = userState.get("A1");
+    userState.clear();
+    assertArrayEquals(new String[] {"V1", "V2"}, Iterables.toArray(initValues, String.class));
+    assertTrue(Iterables.isEmpty(userState.get("A1")));
```

Comment:
nit: rely on `assertThat(collection, IsEmpty());`

Makes for better error messages when tests fail.